File: chat/common/js/src/main/resources/util/benchmark.js
$(function() {
  "use strict"

  window.Benchmark = function(
      nameChanged, chatRequested, chatSelected, chatClosed, messageSent) {
    var self = this
    self.storeLog = false

    self.name = "Anonymous"

    var State = {
      Initial: 0,
      Waiting: 1,
      Requesting: 2,
      Configuring: 3,
      Pinging: 4,
      Ponging: 5
    }

    var state = State.Initial
    var totalMessagesPerIteration = 0
    var totalIterations = 0
    var warmupIterations = 0

    var iteration = 0
    var count = 0
    var pingMessage = true
    var pongMessage = false
    var time = 0

    var results = null

    self.updateUsers = function(users) {
      if (state == State.Initial) {
        if (users.length == 0) {
          state = State.Waiting
          ui.username.val("Ponging")
          ui.log("[Benchmark mode]")
          ui.log("Waiting for chat request ...")
        }
        else if (users.length == 1) {
          state = State.Requesting
          ui.username.val("Pinging")
          ui.log("[Benchmark mode]")
          ui.log("Requesting chat ...")
          chatRequested(users[0])
        }
        else
          ui.log("[Benchmark mode]: Two users already connected")
      }
    }

    self.updateChats = function(chats) {
      if (chats.length == 1) {
        if (state == State.Requesting) {
          state = State.Configuring
          ui.log("Please enter the message and iteration count in the message field")
          ui.message.val("[message count] [iteration count]")
          ui.message.prop("disabled", false)
          ui.send.prop("disabled", false)
          chatSelected(chats[0])
        }
        else if (state == State.Waiting) {
          state = State.Ponging
          ui.log("Running benchmark ...")
          chatSelected(chats[0])
        }
      }
    }

    self.updateMessages = function(messages) {
      if (messages instanceof Cons) {
        if (state == State.Pinging) {
          if (pingMessage = !pingMessage) {
            count++
            if (count >= totalMessagesPerIteration) {
              iteration++

              if (iteration < totalIterations) {
                if (iteration >= 0)
                  results[iteration] = ((1000 * performance.now()) | 0) - time

                count = 0
                setTimeout(function() { ping() }, 1)
              }
              else if (iteration == totalIterations) {
                var min = (Math.min.apply(null, results) / totalMessagesPerIteration) | 0
                var max = (Math.max.apply(null, results) / totalMessagesPerIteration) | 0
                var mean = results.reduce(function(a, b) { return a + b }) / results.length / totalMessagesPerIteration
                var variance = results.reduce(function(variance, element) {
                  var diff = element / totalMessagesPerIteration - mean
                  return variance + diff * diff
                }, 0) / (results.length - 1)
                var standardDeviation = Math.sqrt(variance)
                var standardErrorMean = standardDeviation / Math.sqrt(results.length)

                ui.log("MIN: " + min + "μs")
                ui.log("MAX: " + max + "μs")
                ui.log("AVG: " + (mean | 0) + "μs")
                ui.log("SEM: " + (standardErrorMean | 0) + "μs")
                ui.log("SD:  " + (standardDeviation | 0) + "μs")
              }
            }
          }
        }
        else if (state == State.Ponging) {
          if (pongMessage = !pongMessage)
            messageSent("pong for " + messages.head.content)
        }
      }
    }

    function ping() {
      time = (1000 * performance.now()) | 0
      for (var i = 0; i < totalMessagesPerIteration; i++)
        messageSent("ping " + i)
    }

    self.clearMessage = function() { }

    var ui = {
      username: $("#username"),
      chats: $("#chats"),
      users: $("#users"),
      nousers: $("#nousers"),
      chatlog: $("#chatlog"),
      message: $("#message"),
      send: $("#send"),
      log: function(message) {
        ui.chatlog.append($("<li/>").text(message))
      }
    }

    ui.message.attr("placeholder", "")
    ui.message.prop("disabled", true)
    ui.send.prop("disabled", true)
    ui.username.prop("disabled", true)

    ui.send.on("click", function() {
      var matched = ui.message.val().match(/^\s*(\d+)\s+(\d+)\s*$/)

      if (matched != null) {
        totalMessagesPerIteration = +matched[1]
        totalIterations = +matched[2]
        warmupIterations = Math.max(5, (5000 / totalMessagesPerIteration))
        iteration = -warmupIterations

        results = new Array(totalIterations)

        ui.message.val("")
        ui.message.prop("disabled", true)
        ui.send.prop("disabled", true)

        ui.log(
          totalIterations + " iterations with " +
          totalMessagesPerIteration + " messages each " +
          "(plus " + warmupIterations + " warm-up iterations)")

        ui.log("Running benchmark ...")

        state = State.Pinging

        setTimeout(function() { ping() }, 1)
      }
      else
        alert("Cannot parse message and iteration count")
    })

    ui.message.on("keyup", function(event) {
      if (event.keyCode == 13) {
        event.preventDefault()
        ui.send.trigger("click")
      }
    })
  }
})

File: chat/common/scalajs/src/main/scala/chat/common/Benchmark.scala
package chat
package common

import util._

import scala.scalajs.js.timers
import scala.scalajs.js.Dynamic
import scala.scalajs.js.Dynamic.global

class Benchmark(
    chatRequested: User => Unit,
    chatSelected: Chat => Unit,
    messageSent: String => Unit) {
  sealed trait State
  case object Initial extends State
  case object Waiting extends State
  case object Requesting extends State
  case object Configuring extends State
  case object Pinging extends State
  case object Ponging extends State

  private[this] var state: State = Initial
  private[this] var totalMessagesPerIteration: Int = _
  private[this] var totalIterations: Int = _
  private[this] var warmupIterations: Int = _

  private[this] var iteration: Int = _
  private[this] var count: Int = _
  private[this] var pingMessage: Boolean = true
  private[this] var pongMessage: Boolean = false
  private[this] var time: Long = _

  private[this] var results: Array[Long] = _

  private object ui {
    var username: Dynamic = _
    var chats: Dynamic = _
    var users: Dynamic = _
    var nousers: Dynamic = _
    var chatlog: Dynamic = _
    var message: Dynamic = _
    var send: Dynamic = _
    def log(message: String) =
      chatlog.append($("<li/>").text(message))
  }

  private val $ = global.$

  def updateUsers(users: Seq[User]): Unit = $ { () =>
    if (state == Initial) {
      if (users.size == 0) {
        state = Waiting
        ui.username.`val`("Ponging")
        ui.log("[Benchmark mode]")
        ui.log("Waiting for chat request ...")
      }
      else if (users.size == 1) {
        state = Requesting
        ui.username.`val`("Pinging")
        ui.log("[Benchmark mode]")
        ui.log("Requesting chat ...")
        chatRequested(users.head)
      }
      else
        ui.log("[Benchmark mode]: Two users already connected")
    }
  }

  def updateChats(chats: Seq[Chat]): Unit = $ { () =>
    if (chats.size == 1) {
      if (state == Requesting) {
        state = Configuring
        ui.log("Please enter the message and iteration count in the message field")
        ui.message.`val`("[message count] [iteration count]")
        ui.message.prop("disabled", false)
        ui.send.prop("disabled", false)
        chatSelected(chats.head)
      }
      else if (state == Waiting) {
        state = Ponging
        ui.log("Running benchmark ...")
        chatSelected(chats.head)
      }
    }
  }

  def updateMessages(messages: Seq[Message]): Unit = $ { () =>
    if (messages.nonEmpty) {
      if (state == Pinging) {
        pingMessage = !pingMessage
        if (pingMessage) {
          count += 1
          if (count >= totalMessagesPerIteration) {
            iteration += 1

            if (iteration < totalIterations) {
              if (iteration >= 0)
                results(iteration) = System.nanoTime - time

              count = 0
              timers.setTimeout(1) { ping() }
            }
            else if (iteration == totalIterations) {
              val min = results.min / totalMessagesPerIteration / 1000
              val max = results.max / totalMessagesPerIteration / 1000
              val mean = results.sum.toDouble / results.size / totalMessagesPerIteration / 1000
              val variance = (results.foldLeft(0.0) { (variance, element) =>
                val diff = element / totalMessagesPerIteration / 1000 - mean
                variance + diff * diff
              }) / (results.size - 1)
              val standardDeviation = math.sqrt(variance)
              val standardErrorMean = standardDeviation / math.sqrt(results.size)

              ui.log(s"MIN: ${min}μs")
              ui.log(s"MAX: ${max}μs")
              ui.log(s"AVG: ${mean.toLong}μs")
              ui.log(s"SEM: ${standardErrorMean.toLong}μs")
              ui.log(s"SD:  ${standardDeviation.toLong}μs")
            }
          }
        }
      }
      else if (state == Ponging) {
        pongMessage = !pongMessage
        if (pongMessage)
          messageSent(s"pong for ${messages.head.content}")
      }
    }
  }

  private def ping(): Unit = {
    time = System.nanoTime
    for (i <- 1 to totalMessagesPerIteration)
      messageSent(s"ping $i")
  }

  $ { () =>
    ui.username = global $ "#username"
    ui.chats = global $ "#chats"
    ui.users = global $ "#users"
    ui.nousers = global $ "#nousers"
    ui.chatlog = global $ "#chatlog"
    ui.message = global $ "#message"
    ui.send = global $ "#send"

    ui.username.`val`("")
    ui.message.`val`("")
    ui.message.attr("placeholder", "")
    ui.message.prop("disabled", true)
    ui.send.prop("disabled", true)
    ui.username.prop("disabled", true)

    ui.send.on("click", { () =>
      val userInput = """\s*(\d+)\s+(\d+)\s*""".r

      ui.message.`val`().toString match {
        case userInput(messages, iterations) =>
          totalMessagesPerIteration = messages.toInt
          totalIterations = iterations.toInt
          warmupIterations = math.max(5, (5000 / totalMessagesPerIteration))
          iteration = -warmupIterations

          results = new Array(totalIterations)

          ui.message.`val`("")
          ui.message.prop("disabled", true)
          ui.send.prop("disabled", true)

          ui.log(
            s"$totalIterations iterations with " +
            s"$totalMessagesPerIteration messages each " +
            s"(plus $warmupIterations warm-up iterations)")

          ui.log("Running benchmark ...")

          state = Pinging

          timers.setTimeout(1) { ping() }

        case _ =>
          global.alert("Cannot parse message and iteration count")
      }
    })

    ui.message.on("keyup", { event: Dynamic =>
      if (event.keyCode.asInstanceOf[Int] == 13) {
        event.preventDefault()
        ui.send trigger "click"
      }
    })
  }
}
